components/raftstore/src/store/fsm/apply.rs

```diff
@@ -2194,18 +2198,19 @@ where
                     let (role, exist_id, incoming_id) =
                         (exist_peer.get_role(), exist_peer.get_id(), peer.get_id());
 
+                    let is_witness_change = exist_peer.is_witness != peer.is_witness;
                     if exist_id != incoming_id // Add peer with different id to the same store
                             // The peer is already the requested role
-                            || (role, change_type) == (PeerRole::Voter, ConfChangeType::AddNode)
-                            || (role, change_type) == (PeerRole::Learner, ConfChangeType::AddLearnerNode)
+                            || ((role, change_type) == (PeerRole::Voter, ConfChangeType::AddNode) && !is_witness_change)
```

Comment:
What could go wrong if `add_witness` is added as a new command instead of mixing with conf change?

Because changing between witness and non-witness have nothing to do with configuration change, and doesn't even need to utilize joint consensus. Using a separate new command seems more clear to me.

Reply:
But there is nothing to be changed in conf change function, right? Only those conversions between witness and non-witness go to other command.

Reply:
I agree it's clearer. But the change in the conf change function is small, using a separate command would introduce a lot of changes. Not benefit shows that it is worth doing that.

Reply:
> a lot of changes

Why? It should be less than 10 lines of code in apply.

```
fn apply_switch_witness(&self, changes) {
    for peer in changes {
        for p in self.region_mut().mut_peers() {
            if p.id == peer.id {
                *p = peer;
                break;
            }
        }
    }
}
```

Reply:
need to change all along the path of PD dispatching operator, and the epoch should be increased as well.

Reply:
> Need to change all along the path of PD dispatching operator

Should it just a function? The algorithm is the same, the only difference is dispatching conf change or the new admin command. Besides, how to tell if a conversion is finished is also different from conf change on PD side. PD needs to check if it's in wait data list. And with a new command, you don't even need to enforce all conf changes have to be done by v2.

Note the change also needs to be ported to v2, using a new command can also make the process less headache.

Reply:
> Should it just a function? The algorithm is the same, the only difference is dispatching conf change or the new admin command. Besides, how to tell if a conversion is finished is also different from conf change on PD side. PD needs to check if it's in wait data list. 

Actually, it's simple to change on PD side, the structure is extensible enough. What I mean is the process on TiKV side when receiving a new kind of operator and checks when proposing. It does not hard to change, but still needs more effort than changing conf change directly. Though I agree using a new command can also make the process less headache when porting to v2.

Will remove the support of switching witness by conf change in this PR, and use a new command to support in the next PR. Please move on to review the rest of the code.

Reply:
> And with a new command, you don't even need to enforce all conf changes have to be done by v2.
BTW, it's already been done, all conf changes are handled by change_peer_v2.

Reply:
/cc @ethercflow